pkg/proxy/handlers/spacelister.go

```diff
@@ -147,7 +147,10 @@ func (s *SpaceLister) GetUserWorkspace(ctx echo.Context) (*toolchainv1alpha1.Wor
 	), nil
 }
 
-func (s *SpaceLister) ListUserWorkspaces(ctx echo.Context) ([]toolchainv1alpha1.Workspace, error) {
+// ListUserWorkspaces returns a list of Workspaces for the current user.
+// If a workspaceName is provided, it means the request was made for a specific workspace, the function will check if the user has access to that workspace, if not it will return 404 Unauthorized.
+// If no workspaceName is provided (workspaceName is an empty sting), the function will list all SpaceBindings for the user and return all the workspaces found from this list.
+func (s *SpaceLister) ListUserWorkspaces(ctx echo.Context, workspaceName string) ([]toolchainv1alpha1.Workspace, error) {
```

Comment:
TBH, it took me a while to understand the usage and the cases when the method is being used. It's most likely because it's mixing two  use-cases:
1. Get list of workspaces the user has access to - this is done via listing the SpaceBindings by the username
2. Validate access for the given workspace and get information about it - but this is done by collecting all SpaceBindings for the given space and all parent-spaces

in other words - two different solutions:
1. getting all (inherited) SpaceBindings of the Space and filtering them
2. getting all SpaceBindings of the User

at the same time there is also the other case that collects all SpaceBindings for the Space: https://github.com/codeready-toolchain/registration-service/blob/3d66cb33031e62543a58332f27708fbf1c5e9394/pkg/proxy/handlers/spacelister.go#L108-L128

I'm wondering if there is a better way to use/organize the code to easily follow the use-cases.

Reply:
You're right! the code is a bit "messy". I've done some refactor, attempting to remove some of the duplication, and make those paths more clear.

In https://github.com/codeready-toolchain/registration-service/pull/371/commits/796bed8496eb74d47f5ec2a1896cf47a85c87c02#diff-80be0a15eaa87db8f407cfa1ecfa516691ddee93b5fedb1cebc3347fec7801c8 I've:
- introduced `GetUserWorkspace` also in the proxy handler (instead of the new function which basically was doing the same thing, as you pointed out)
- moved the hierarchical unit tests to the `HandleSpaceGetRequest` subtest 
- splitted the get and list handlers functions from `spacelister` ( this was not strictly required and I can revert it, but it makes the code a bit easier to navigate IMO)

I'm planning also to split the `spacelister_test.go` into 2 separate tests as I've added into the PR description, if this makes sense. 

Please let me know WDYT.